src/views/__tests__/NavigationEvents-test.js

```diff
@@ -0,0 +1,164 @@
+import { View } from 'react-native';
+import renderer from 'react-test-renderer';
+import NavigationEvents from '../NavigationEvents';
+import { NavigationProvider } from '../NavigationContext';
+
+describe('NavigationEvents', () => {
+  let onWillFocus;
+  let onDidFocus;
+  let onWillBlur;
+  let onDidBlur;
+  let navigation;
+  let checkAddListenerCalled;
+  let checkAddListenerCalledWith;
+  let TestComponent;
+  let TestComponentNoContext;
+
+  // A little API to spy on subscription remove calls
+  let RemoveCallsAPI = (() => {
+    let removeCalls = [];
+    return {
+      reset: () => {
+        removeCalls = [];
+      },
+      add: (name, handler) => {
+        removeCalls.push({ name, handler });
+      },
+      expectCallNumber: count => {
+        expect(removeCalls.length).toBe(count);
+      },
+      expectCall: (name, handler) => {
+        expect(removeCalls).toContainEqual({ name, handler });
+      },
+    };
+  })();
+
+  const createListener = () => payload => {};
+  const createNavigation = () => {
+    return {
+      addListener: jest.fn((name, handler) => {
+        return {
+          remove: () => RemoveCallsAPI.add(name, handler),
+        };
+      }),
+    };
+  };
+
+  beforeEach(() => {
+    RemoveCallsAPI.reset();
+    onWillFocus = createListener();
+    onDidFocus = createListener();
+    onWillBlur = createListener();
+    onDidBlur = createListener();
+    navigation = createNavigation();
+    checkAddListenerCalled = count => {
+      expect(navigation.addListener).toHaveBeenCalledTimes(count);
+    };
+    checkAddListenerCalledWith = (eventName, handler) => {
+      expect(navigation.addListener).toHaveBeenCalledWith(eventName, handler);
+    };
+
+    // eslint-disable-next-line react/display-name
+    TestComponent = props => (
+      <NavigationProvider value={navigation}>
+        <NavigationEvents
+          onWillFocus={onWillFocus}
+          onDidFocus={onDidFocus}
+          onWillBlur={onWillBlur}
+          onDidBlur={onDidBlur}
+          {...props}
+        />
+      </NavigationProvider>
+    );
+
+    // eslint-disable-next-line react/display-name
+    TestComponentNoContext = props => (
+      <View>
+        <NavigationEvents
+          onWillFocus={onWillFocus}
+          onDidFocus={onDidFocus}
+          onWillBlur={onWillBlur}
+          onDidBlur={onDidBlur}
+          {...props}
+        />
+      </View>
+    );
+  });
+
+  it('attach all listeners with navigation prop', () => {
+    const component = renderer.create(
+      <TestComponentNoContext navigation={navigation} />
+    );
+    checkAddListenerCalled(4);
+    checkAddListenerCalledWith('willBlur', onWillBlur);
+    checkAddListenerCalledWith('willFocus', onWillFocus);
+    checkAddListenerCalledWith('didBlur', onDidBlur);
+    checkAddListenerCalledWith('didFocus', onDidFocus);
+  });
+
+  it('attach all listeners with navigation context', () => {
+    const component = renderer.create(<TestComponent />);
```

Comment:
these tests seem to do the job, but personally I'm a fan of avoiding much work in `beforeEach` and instead making it so that each test reads more like a standalone story. The way you have it here, I need to read the whole complicated test harness to understand a single test. This results in longer test files, but I think the improved readability makes it worthwhile.

I'd say its fine for now but in the future you may want to keep the whole test isolated to within the `it(` callback